File: pfizer_audit_logger.py
#!/usr/bin/env python3
#
# Tested with python 3.4
#
# This script will read json-encoded auditlog messages from Transmart and send them as HTTP POST messages to
# metrics.pfizer.com.

import sys
import json
import concurrent.futures
import urllib.request
import urllib.parse
from urllib.request import Request
# httpagentparser is not in the standard library, use "pip3 install httpagentparser" to install it
import httpagentparser

url = 'http://metrics.pfizer.com/metrics/servlet/RegisterEventServlet?'
TIMEOUT = 4 #seconds


failcount = 0

def send_url(msg):
    global failcount
    
    action = [msg.get('action')] or [msg.get('study'), msg.get('subset1'), msg.get('subset2')]
    action = '|'.join([i for i in action if i])
    args = dict(action = action,
                application = msg['program'],
                appVersion = msg['programVersion'],
                user = msg['user'],
                task = msg['event'],
                browser = httpagentparser.detect(msg['userAgent'])['browser']['name']
            )
    fullurl = url + urllib.parse.urlencode(args)
    #print(fullurl)
    try:
        urllib.request.urlopen(Request(fullurl, method='POST'), timeout=TIMEOUT).readall()
        failcount = 0
    except Exception as e:
        failcount += 1
        if failcount > 15:
            raise
        print(e, file=sys.stderr)


with concurrent.futures.ThreadPoolExecutor(max_workers=20) as executor:
    for line in sys.stdin:
        msg = json.loads(line)
        send_url(msg)

sys.stdin.close()

File: pfizer_audit_logger.go
package main

import (
	"os"
	"time"
	"sync"
	"sync/atomic"
	"encoding/json"
	"strings"
	"net/url"
	"net/http"
	"io/ioutil"
	//"log"
	//"reflect"
	"github.com/mssola/user_agent"
)

import "fmt"

const URL = "http://metrics.pfizer.com/metrics/servlet/RegisterEventServlet"

// If there are more than this number of unprocessed messages, further messages will be dropped until the server
// catches up.
const MAX_QUEUE_LENGTH = 1000


var wg sync.WaitGroup
// The internal count of WaitGroup is not accessible, so we maintain another copy to know how many requests are in progress
var requests_in_progress int32 = 0

func atomicAdd(i *int32, delta int32) { atomic.AddInt32(i, delta) }
func atomicLoad(i *int32) int32 { return atomic.LoadInt32(i) }


//type jsonMsg map[string]interface{}

type jsonMsg struct {
	Action string
	Study string
	Subset1 string
	Subset2 string
	Application string `json:"program"`
	AppVersion string `json:"programVersion"`
	User string
	Task string `json:"event"`
	Browser string `json:"userAgent"`
}


func removeEmpty(in []string) []string {
	res := []string{}
	for _, i := range in {
		if i != "" {
			res = append(res, i)
		}
	}
	return res
}

func send_auditlog_record(msg jsonMsg) {

	//fmt.Println(msg)
	params := url.Values{}

	if msg.Action != "" {
		params.Set("action", msg.Action)
	} else {
		action := []string{}
		for _, s := range []string{msg.Study, msg.Subset1, msg.Subset2} {
			if s != "0" && s != "" {
				action = append(action, s)
			}
		}
		params.Set("action", strings.Join(action, "|"))
	}
	if msg.Application != "" { params.Set("application", msg.Application) }
	if msg.AppVersion != "" { params.Set("appVersion", msg.AppVersion) }
	if msg.User != "" { params.Set("user", msg.User) }
	if msg.Task != "" { params.Set("task", msg.Task) }
	if msg.Browser != "" {
		ua := user_agent.New(msg.Browser)
		browser, version := ua.Browser()
		params.Set("browser", browser + " " + version)
	}

//	for k, v := range msg {
//		params.Set(k, fmt.Sprintf("%v", v))
//	}

	fullurl := URL + "?" + params.Encode()

	fmt.Println(fullurl)
	
	resp, err := http.PostForm(fullurl, url.Values{})
	if err != nil {
		error("WARNING: Failed to post message to metrics server: "+err.Error())
	} else {
		defer resp.Body.Close()
		ioutil.ReadAll(resp.Body)
	}

	atomicAdd(&requests_in_progress, -1)
	wg.Done()
}


func error(m string) {
	fmt.Fprintln(os.Stderr, time.Now().Format("2006-01-02 15:04:05.000-07:00  ") + m)
}

func fatal(m string) {
	error(m)
	os.Exit(1)
}


func main() {
	
	dec := json.NewDecoder(os.Stdin)
	msg := jsonMsg{}
	msgp := &msg

	for dec.More() {
		err := dec.Decode(&msgp)
		if err != nil { fatal("ERROR: Failed to parse JSON message from stdin: " + err.Error()) }
		if atomicLoad(&requests_in_progress) > MAX_QUEUE_LENGTH {
			line, _ := json.Marshal(msg)
			error("WARNING: Too many unfinished requests in progress, ignoring message "+string(line))
			continue
		}
		atomicAdd(&requests_in_progress, 1)
		wg.Add(1)
		go send_auditlog_record(msg)
		msg = jsonMsg{}
	}

	//fmt.Println("done reading")
	wg.Wait()
	//fmt.Println("Exiting")
}

